winch/codegen/src/isa/aarch64/address.rs

```diff
@@ -0,0 +1,142 @@
+//! Aarch64 addressing mode.
+
+use anyhow::{anyhow, Context, Result};
+use cranelift_codegen::{
+    ir::types,
+    isa::aarch64::inst::{AMode, PairAMode, SImm7Scaled, SImm9},
+};
+
+use super::regs;
+use crate::reg::Reg;
+
+/// Aarch64 indexing mode.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub(crate) enum Indexing {
+    /// Pre-indexed.
+    Pre,
+    /// Post-indexed.
+    Post,
+}
+
+/// Memory address representation.
+#[derive(Debug, Copy, Clone)]
+pub(crate) enum Address {
+    /// Base register with an arbitrary offset.  Potentially gets
+    /// lowered into multiple instructions during code emission
+    /// depending on the offset.
+    Offset {
+        /// Base register.
+        base: Reg,
+        /// Offset.
+        offset: i64,
+    },
+    /// Specialized register and offset variant using
+    /// the stack pointer.
+    SPOffset {
```

Comment:
Since this variant only supports pre-/post-indexed forms (i.e., forms that update `SP`), maybe we can name it `SPOffsetWithUpdate` (or something less cumbersome, though nothing comes to mind at the moment...)?

Reply:
I renamed this to `IndexedSPOffset`, let me know if that's clear enough. Else I'm happy to find a better name.